### PR TITLE
Explosive Shot

### PR DESCRIPTION
**Pull request by AntonAsbrink**
Ranged Attack causes shrapnel to eject from enemies hit damaging nearby enemies for a precentage of the damage

### CODE REVIEW

**Review by jocke1995 — DISMISSED** [review]
good job

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	m_Level++;
+	m_Radius = (5 * m_Level);
+	m_PercentageDamage = 0.4 + (0.1 * m_Level);
+	m_Price += m_StartingPrice;
+}
+
+void UpgradeExplosiveAttack::OnRangedHit(Entity* target)
+{
+	//Should only affect enemies. Only enemies should have AiComponent and therefore should be a good test
+	if (target->HasComponent<component::AiComponent>())
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: This hasComponent is a very slow function, if there is a better way to check for enemy, use it.

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+		ParticleEffectSettings settings = {};
+		settings.maxParticleCount = 20;
+		settings.startValues.lifetime = 1.0;
+		settings.spawnInterval = 0.001;
+		settings.startValues.acceleration = { 0, 0, 0 };
+		settings.isLooping = false;
+
+		// Need to fix EngineRand.rand() for negative values
+		RandomParameter3 randParam1 = { -m_Radius, m_Radius, -m_Radius, m_Radius, -m_Radius, m_Radius };
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: Have you somehow debugged the visual radius of the actual radius?

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+					enemies->at(i)->GetComponent<component::HealthComponent>()->TakeDamage(damage);
+				}
+			}
+		}
+
+		ParticleEffectSettings settings = {};
+		settings.maxParticleCount = 20;
+		settings.startValues.lifetime = 1.0;
+		settings.spawnInterval = 0.001;
+		settings.startValues.acceleration = { 0, 0, 0 };
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: Would create a "base" ParticleEffectSettings on constructor, then in here modify that one with the only things that need to be modified, it is then later copied when creating the particleeffect.

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		settings.isLooping = false;
+
+		// Need to fix EngineRand.rand() for negative values
+		RandomParameter3 randParam1 = { -m_Radius, m_Radius, -m_Radius, m_Radius, -m_Radius, m_Radius };
+
+		settings.randPosition = { -1, 1, -1, 1, -1, 1 };
+		settings.randVelocity = randParam1;
+		settings.randSize = { 0.4, 0.8 };
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: When the radius gets very big, the effect looks weird with such small particle sizes, would like to see this also to be effected by radius

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		settings.randRotationSpeed = { 0, 3 };
+
+		settings.texture = static_cast<Texture2DGUI*>(AssetLoader::Get()->LoadTexture2D(L"../Vendor/Resources/Textures/Particles/shrapnel_particle.png"));
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: Very nice texture ;)

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		int particleCounter = 0;
+		while (SceneManager::GetInstance().GetActiveScene()->EntityExists("explosionParticle" + std::to_string(particleCounter)))
+		{
+			particleCounter++;
+		}
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: Not a good solution looping through each entity on hit, you can create a static unsigned int particleCounter in this class which gets increased.

File: StortSpelProjekt/Engine/src/ECS/Components/ParticleEmitterComponent.h [hunk 7]
```diff
@@ -36,6 +36,7 @@ namespace component
 
         std::vector<ParticleEffect*> m_ParticleEffects;
         bool m_PlayOnInit;
+        double m_PlayDuration;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/ECS/Components/ParticleEmitterComponent.h]: if not used, remove.

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp (hunk 1 above)
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: Game design thing:
Should you not be able to hit the ground and do AoE damage? Feels strange from a player perspective to have explosive shots that only explodes on enemies. Not sure if it's possible in our code but worth thinking about.

**Review by fjz345 — CHANGES_REQUESTED** [review]
comment back on each comment and I'll check again later ;). "fixed" is enough if you just changed simple stuff.

**Review by Schtekt — CHANGES_REQUESTED** [review]
The code looks good and the particles are nice. I don't see how many enemies are hit by my explosive attacks though?
Is it possible to give a hit particle effect of when an enemy is hit by the explosion and not the projectile?

File: StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp (hunk 1 above)
Comment (AntonAsbrink) [reply StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp]: As far as i can see impossible. I can not reliably get the position of collision. Thats why i reflavored it as i did

**Comment by AntonAsbrink** [comment]
> The code looks good and the particles are nice. I don't see how many enemies are hit by my explosive attacks though?
> Is it possible to give a hit particle effect of when an enemy is hit by the explosion and not the projectile?

I don't think i should or will make that. We really should have a general feedback for when enemies get damaged

**Review by jocke1995 — DISMISSED** [review]
Works perfectly!

**Review by jocke1995 — DISMISSED** [review]

**Review by jocke1995 — APPROVED** [review]

**Review by fjz345 — APPROVED** [review]

**Review by Schtekt — APPROVED** [review]

**Pull request merged**